//===----------------------------------------------------------------------===//
//
// This source file is part of the Teco open source project.
//
// Copyright (c) 2022 the Teco project authors
// Licensed under Apache License v2.0
//
// See LICENSE.txt for license information
//
// SPDX-License-Identifier: Apache-2.0
//
//===----------------------------------------------------------------------===//

// THIS FILE IS AUTOMATICALLY GENERATED by TecoServiceGenerator.
// DO NOT EDIT.

extension TCLighthouseError {
    public struct LimitExceeded: TCErrorType {
        enum Code: String {
            case attachDataDiskQuotaLimitExceeded = "LimitExceeded.AttachDataDiskQuotaLimitExceeded"
            case firewallRulesLimitExceeded = "LimitExceeded.FirewallRulesLimitExceeded"
            case instanceQuotaLimitExceeded = "LimitExceeded.InstanceQuotaLimitExceeded"
            case isolateResourcesLimitExceeded = "LimitExceeded.IsolateResourcesLimitExceeded"
            case keyPairLimitExceeded = "LimitExceeded.KeyPairLimitExceeded"
            case snapshotQuotaLimitExceeded = "LimitExceeded.SnapshotQuotaLimitExceeded"
            case other = "LimitExceeded"
        }
        
        private let error: Code
        
        public let context: TCErrorContext?
        
        public var errorCode: String {
            self.error.rawValue
        }
        
        /// Initializer used by ``TCClient`` to match an error of this type.
        ///
        /// You should not use this initializer directly as there are no public initializers for ``TCErrorContext``.
        public init ?(errorCode: String, context: TCErrorContext) {
            guard let error = Code(rawValue: errorCode) else {
                return nil
            }
            self.error = error
            self.context = context
        }
        
        internal init (_ error: Code, context: TCErrorContext? = nil) {
            self.error = error
            self.context = context
        }
        
        /// 实例挂载数据盘配额不足，无法挂载磁盘。
        public static var attachDataDiskQuotaLimitExceeded: LimitExceeded {
            LimitExceeded(.attachDataDiskQuotaLimitExceeded)
        }
        
        /// 超过防火墙规则配额。
        public static var firewallRulesLimitExceeded: LimitExceeded {
            LimitExceeded(.firewallRulesLimitExceeded)
        }
        
        /// 超过实例配额。
        public static var instanceQuotaLimitExceeded: LimitExceeded {
            LimitExceeded(.instanceQuotaLimitExceeded)
        }
        
        /// 退还资源数量超出限制。
        public static var isolateResourcesLimitExceeded: LimitExceeded {
            LimitExceeded(.isolateResourcesLimitExceeded)
        }
        
        /// 超过密钥对配额。
        public static var keyPairLimitExceeded: LimitExceeded {
            LimitExceeded(.keyPairLimitExceeded)
        }
        
        /// 超过快照配额。
        public static var snapshotQuotaLimitExceeded: LimitExceeded {
            LimitExceeded(.snapshotQuotaLimitExceeded)
        }
        
        /// 超过配额限制。
        public static var other: LimitExceeded {
            LimitExceeded(.other)
        }
    }
}

extension TCLighthouseError.LimitExceeded: Equatable {
    public static func == (lhs: TCLighthouseError.LimitExceeded, rhs: TCLighthouseError.LimitExceeded) -> Bool {
        lhs.error == rhs.error
    }
}

extension TCLighthouseError.LimitExceeded: CustomStringConvertible {
    public var description: String {
        return "\(self.error.rawValue): \(message ?? "")"
    }
}

extension TCLighthouseError.LimitExceeded {
    /// - Returns: ``TCLighthouseError`` that holds the same error and context.
    public func toLighthouseError() -> TCLighthouseError {
        guard let code = TCLighthouseError.Code(rawValue: self.error.rawValue) else {
            fatalError("Unexpected internal conversion error!\nPlease file a bug at https://github.com/teco-project/teco to help address the problem.")
        }
        return TCLighthouseError(code, context: self.context)
    }
}

extension TCLighthouseError.LimitExceeded {
    /// - Returns: ``TCCommonError`` that holds the same error and context.
    public func toCommonError() -> TCCommonError? {
        if let context = self.context, let error = TCCommonError(errorCode: self.error.rawValue, context: context) {
            return error
        }
        return nil
    }
}
